### PR TITLE
fix(ilp): fix variability in ILP performance

### PR DESCRIPTION
**Pull request by bluestreak01**
Together with https://github.com/questdb/questdb/pull/2635 this PR stabilized performance of ILP ingestion by eliminating 
[megamorphic virtual calls](https://shipilev.net/jvm/anatomy-quarks/16-megamorphic-virtual-calls/). Specifically `charAt()` calls on `CharSequence` implementations.

On my desktop I get very stable TSBS performance of 1.2+M rows/s. Never below (notwithstanding short default lag, which could send TSBS into O3)

### CODE REVIEW

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 130 / 131 (99.24%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/std/DirectByteCharSequenceIntHashMap.java|120|121|99.17%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/TableUpdateDetails.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/std/Chars.java|8|8|100.00%|

File: core/src/main/java/io/questdb/std/DirectByteCharSequenceIntHashMap.java [hunk 1]
```diff
@@ -0,0 +1,279 @@
...
+ ******************************************************************************/
+
+package io.questdb.std;
+
+import io.questdb.std.str.DirectByteCharSequence;
+
+import java.util.Arrays;
+
+
+public class DirectByteCharSequenceIntHashMap implements Mutable {
```
Comment (ideoma) [inline core/src/main/java/io/questdb/std/DirectByteCharSequenceIntHashMap.java:32]: nit: a comment explaining why this class is created would be good to have

**Review by ideoma — APPROVED** [review]

**Review by puzpuzpuz — COMMENTED** [review]
I can still see some jitter on my Ubuntu 22.04 x86-64 machine: 420K-620K rows/s for ingestion a single ILP connection, but the improvement is indeed very noticeable. 620K rows/s on a single connection is very impressive. 👍

**Comment by bluestreak01** [comment]
@puzpuzpuz could you try async-profiler on the ubuntu ? I tried eliminating all of the `itable` instances. There should be none. It could be something else this time

File: core/src/main/java/io/questdb/std/DirectByteCharSequenceIntHashMap.java [hunk 2]
```diff
@@ -0,0 +1,279 @@
...
+    public DirectByteCharSequenceIntHashMap(int initialCapacity, double loadFactor, int noEntryValue) {
+        if (loadFactor <= 0d || loadFactor >= 1d) {
+            throw new IllegalArgumentException("0 < loadFactor < 1");
+        }
+
+        this.free = this.capacity = initialCapacity < MIN_INITIAL_CAPACITY ? MIN_INITIAL_CAPACITY : Numbers.ceilPow2(initialCapacity);
+        this.loadFactor = loadFactor;
+        int len = Numbers.ceilPow2((int) (this.capacity / loadFactor));
+        this.keys = new String[len];
+        DirectByteCharSequenceIntHashMap.this.mask = len - 1;
```
[the remark this answers is not in the document]
Comment (bluestreak01) [reply core/src/main/java/io/questdb/std/DirectByteCharSequenceIntHashMap.java]: done in https://github.com/questdb/questdb/pull/2654/commits/0ed0196df7322ede907c6865a1d7663af341973f

File: core/src/main/java/io/questdb/std/DirectByteCharSequenceIntHashMap.java (hunk 1 above)
Comment (bluestreak01) [reply core/src/main/java/io/questdb/std/DirectByteCharSequenceIntHashMap.java:40]: done in https://github.com/questdb/questdb/pull/2654/commits/0ed0196df7322ede907c6865a1d7663af341973f

**Comment by ideoma** [comment]
#### [PR Coverage check]

:heart_eyes: **pass** : 130 / 131 (99.24%)




#### file detail

|   |path|covered line|new line|coverage|
|----|----|----|----|----|
|:large_blue_circle:|io/questdb/std/DirectByteCharSequenceIntHashMap.java|120|121|99.17%|
|:large_blue_circle:|io/questdb/cutlass/line/tcp/TableUpdateDetails.java|2|2|100.00%|
|:large_blue_circle:|io/questdb/std/Chars.java|8|8|100.00%|

**Review by puzpuzpuz — APPROVED** [review]

**Comment by puzpuzpuz** [comment]
> @puzpuzpuz could you try async-profiler on the ubuntu ? I tried eliminating all of the `itable` instances. There should be none. It could be something else this time

I did that already - `itable` are gone and the flame graphs look very similar. I tried to check if that could have something to do with false sharing with `perf c2c`, but didn't see any evidence of false sharing.

**Comment by bluestreak01** [comment]
excellent, thank you!

**Pull request merged**